luda-editor/new-client/src/asset_manage_page.rs

```diff
@@ -0,0 +1,165 @@
+    }
+}
+
+struct SelectedAssetFile {
+    name: String,
+    bytes: Vec<u8>,
+}
+async fn select_asset_file() -> Result<SelectedAssetFile> {
+    let (data_tx, mut data_rx) = tokio::sync::mpsc::unbounded_channel();
+
+    let js_handle = namui::wasi::insert_js(
+        include_str!("select_asset_file.js"),
+        Some(move |data: &[u8]| {
+            data_tx.send(data.to_vec()).unwrap();
+        }),
+    )
+    .await;
+
+    // See protocol in select_asset_file.js
+    let name = try_read_file_name(&mut data_rx).await?;
+    let bytes = try_read_file_bytes(&mut data_rx).await?;
+
+    drop(js_handle);
+
+    return Ok(SelectedAssetFile { name, bytes });
+
+    async fn try_read_i32(rx: &mut UnboundedReceiver<Vec<u8>>) -> Result<i32> {
+        let bytes = rx.recv().await.ok_or(anyhow!("data channel closed"))?;
+        if bytes.len() != 4 {
+            return Err(anyhow!("invalid i32 bytes length: {}", bytes.len()));
+        }
+        let mut i32_bytes: [u8; 4] = [0; 4];
+        i32_bytes.copy_from_slice(&bytes);
+        Ok(i32::from_be_bytes(i32_bytes))
+    }
+    async fn try_read_file_name(rx: &mut UnboundedReceiver<Vec<u8>>) -> Result<String> {
+        let name_byte_length = match try_read_i32(rx).await? {
+            x if x > 0 => x,
+            -1 => return Err(anyhow!("file name not selected")),
+            x => return Err(anyhow!("invalid file name length: {x}")),
+        };
+
+        let name_bytes = rx.recv().await.ok_or(anyhow!("data channel closed"))?;
+        if name_bytes.len() != name_byte_length as usize {
+            return Err(anyhow!(
+                "invalid file name bytes length: {}",
+                name_bytes.len()
+            ));
+        }
+        Ok(String::from_utf8(name_bytes)?)
+    }
+    async fn try_read_file_bytes(rx: &mut UnboundedReceiver<Vec<u8>>) -> Result<Vec<u8>> {
+        let file_byte_length = try_read_i32(rx).await?;
+        if file_byte_length < 0 {
+            return Err(anyhow!("invalid file bytes length: {}", file_byte_length));
+        }
+        let mut file_bytes = Vec::with_capacity(file_byte_length as usize);
+        let mut writer = std::io::Cursor::new(&mut file_bytes);
+        let mut read_count = 0;
+
+        while read_count < file_byte_length {
+            let chunk_length = match try_read_i32(rx).await? {
+                x if x > 0 => x,
+                0 => return Err(anyhow!("chunk read aborted")),
+                x => return Err(anyhow!("invalid chunk length: {x}")),
+            };
+            let chunk = rx.recv().await.ok_or(anyhow!("data channel closed"))?;
+            if chunk.len() != chunk_length as usize {
+                return Err(anyhow!(
+                    "invalid chunk bytes length: {}, {} expected",
+                    chunk.len(),
+                    chunk_length
+                ));
+            }
+            read_count += chunk_length;
+            writer.write_all(&chunk)?;
+        }
+        Ok(file_bytes)
+    }
```

Comment:
You don't need this complicated code. just send data from js twice and translate it to string and vec.